File: Heaps/heapsort.py
def maxHeap(arr, i, n):
    left = (2 * i) + 1
    right = (2 * i) + 1
    largest = i
    if left < n and arr[left] > arr[largest]:
        largest = left
    if right < n and arr[right] > arr[largest]:
        largest = right

    if largest != i:
        temp = arr[largest]
        arr[largest] = arr[i]
        arr[i] = temp
        maxHeap(arr, largest, n)

def heapify(arr):
    n = len(arr)

    for i in range(n//2, 0, -1):
        maxHeap(arr, i-1, n)

    return arr

def HeapSort(arr):
    sorted_arr = []
    n = len(arr)

    while len(sorted_arr) < n:
        arr = heapify(arr)
        sorted_arr.append(arr.pop(0))

File: Heaps/heapsort.js
function maxHeap(arr, i) {
    var largest = i
    var left = (2 * i) + 1
    var right = (2 * i) + 2

    if (left < arr.length && arr[left] > arr[largest]){
        largest = left
    }
    if (right < arr.length && arr[right] > arr[largest]) {
        largest = right
    }

    if (largest != i) {
        var temp = arr[i]
        arr[i] = arr[largest]
        arr[largest] = temp
        maxHeap(arr, largest)
    }
}

function heapify(arr) {
    n = arr.length

    for (var i = Math.floor(n / 2); i >= 0; i--){
        maxHeap(arr, i)
    }
    return arr
}

function heapSort(arr) {
    var n = arr.length
    var sorted_arr = []

    while (sorted_arr.length < n) {
        arr = heapify(arr)
        sorted_arr.push(arr[0])
        arr.splice(0, 1);
    }
    return sorted_arr
}
